src/sources/docker.rs

```diff
@@ -0,0 +1,802 @@
+            .unwrap_or(info.created)
+            - 1,
+        timestamps: true,
+        ..Default::default()
+    };
+    let mut stream = docker.logs(&info.id, Some(options));
+    info!(
+        message = "Started listening logs on docker container",
+        id = field::display(&info.id)
+    );
+
+    // Create event streamer
+    let mut state = Some((main, info));
+    let event_stream = tokio::prelude::stream::poll_fn(move || {
+        // !Hot code: from here
+        if let Some(&mut (_, ref mut info)) = state.as_mut() {
+            // Main event loop
+            loop {
+                return match stream.poll() {
+                    Ok(Async::Ready(Some(message))) => {
+                        if let Some(event) = log_to_event(message, info) {
+                            Ok(Async::Ready(Some(event)))
+                        } else {
+                            continue;
+                        }
+                        // !Hot code: to here
+                    }
+                    Ok(Async::Ready(None)) => break,
+                    Ok(Async::NotReady) => Ok(Async::NotReady),
+                    Err(error) => {
+                        error!(message = "docker API container logging error",%error);
+                        Err(())
+                    }
+                };
+            }
+        }
+
+        if let Some((main, info)) = state.take() {
+            // End of stream
+            info!(
+                message = "Stoped listening logs on docker container",
+                id = field::display(&info.id)
+            );
+            // TODO: I am not sure that it's necessary to drive this future to completition
+            tokio::spawn(
+                main.send(info)
+                    .map_err(|e| error!(message="Unable to return ContainerLogInfo to main",%e))
+                    .map(|_| ()),
+            );
+        }
+
+        Ok(Async::Ready(None))
+    })
+    .forward(out.sink_map_err(|_| ()))
+    .map(|_| ());
+
+    // Run event_stream
+    tokio::spawn(event_stream);
+}
+
+/// Expects timestamp at the begining of message
+/// Expects messages to be ordered by timestamps
+fn log_to_event(message: LogOutput, info: &mut ContainerLogInfo) -> Option<Event> {
+    let mut log_event = Event::new_empty_log().into_log();
+
+    let (message, stream) = match message {
+        LogOutput::StdErr { message } => (message, "stderr"),
+        LogOutput::StdOut { message } => (message, "stdout"),
+        _ => return None,
+    };
+    log_event.insert_implicit(event::STREAM.clone(), stream.into());
+
+    let mut splitter = message.splitn(2, char::is_whitespace);
+    let timestamp_str = splitter.next()?;
+    let log = match DateTime::parse_from_rfc3339(timestamp_str) {
+        Ok(timestamp) => {
+            // Timestamp check
+            match info.last_log.as_ref() {
+                // Recieved log has already been processed
+                Some(&(ref last, gen)) => match last.cmp(&timestamp) {
```

Comment:
It is necessary do differentiate all three `cmp` cases, and with this the comparison of `DateTime`-s is done only once. This optimization has value since it is in the hotest code in here.

Reply:
Alright, I'll rewrite it like that.